### PR TITLE
Promise.prototype.finally [fixes #232]

### LINKED ISSUE #232: include proposed `Promise.prototype.finally`
basically just:

``` js
/**
  `finally` will be invoked regardless of the promise's fate just as native
  try/catch/finally behaves
  Synchronous example:

  findAuthor() {
    if (Math.random() > 0.5) {
      throw new Error();
    }
    return new Author();
  }
  try {
    return findAuthor(); // succeed or fail
  } catch(error) {
    return findOtherAuther();
  } finally {
    // always runs
    // doesn't affect the return value
  }

  Asynchronous example:

  findAuthor().catch(function(reason){
    return findOtherAuther();
  }).finally(function(){
    // author was either found, or not
  });

  @method finally
  @param {Function} callback
  @return {Promise}
*/
  finally(callback) {
    let promise = this;
    let constructor = promise.constructor;

    return promise.then(value => constructor.resolve(callback()).then(() => value),
                  reason => constructor.resolve(callback()).then(() => { throw reason; }));
  }
```

Questions:
- should 4.0.0's `es6-promise/auto` polyfill  the whole promise, or just `finally` if `finally` is not present?
- should we wait a-bit longer or  is the risk sufficiently low to just :shipit: cc @domenic

### PR DESCRIPTION
**Pull request by stefanpenner**

### CODE REVIEW

**Comment by stefanpenner** [comment]
@ljharb should I wait until stage 3?

**Comment by ljharb** [comment]
Officially yes, but at this point I do not anticipate changes prior to stage 3 that would affect this PR.

Take that with the appropriate grains of salt :-)

**Comment by eLarocque** [comment]
Hey guys.
Wondering what the update is on this since the finally promise is in stage 3. 
https://github.com/tc39/proposals/blob/master/README.md

There a chance we will see an merge in near future?
Thanks

File: es6-promise.d.ts [hunk 1]
```diff
@@ -34,6 +34,13 @@ export class Promise <R> implements Thenable <R> {
   catch <U> (onRejected?: (error: any) => U | Thenable<U>): Promise<U>;
 
   /**
+   * onSettled is invoked when/if the "promise" settles (either rejects or fulfills);
+   *
+   * @param onSettled called when/if "promise" settles
+   */
+  finally <U> (onSettled?: (callback: any) => U | Thenable<U>): Promise<U>;
```
Comment (ljharb) [inline es6-promise.d.ts]: `onFinally` is the name in the spec

File: lib/es6-promise/promise.js [hunk 2]
```diff
@@ -379,5 +379,51 @@ Promise.prototype = {
...
+  });
+  ```
+
+  @method finally
+  @param {Function} callback
+  @return {Promise}
+*/
+  finally(callback) {
+    let promise = this;
+    let constructor = promise.constructor;
```
Comment (ljharb) [inline lib/es6-promise/promise.js]: this needs to use `Symbol.species` to look up the constructor

**Comment by l1bbcsg** [comment]
Any updates on this? `finally` has hit stage 3 several moths ago now.

Also I'd like to remind you that just like `catch` this creates problems in browsers not compliant with ES5 allowing reserved words to be used as property names.
Assuming build process handles that and will quote it just like `'catch'`, it's worth noting in the readme at least.

File: lib/es6-promise/promise.js (hunk 2 above)
Comment (stefanpenner) [reply lib/es6-promise/promise.js]: We don't currently support species stuff here.

**Comment by stefanpenner** [comment]
> Any updates on this? finally has hit stage 3 several moths ago now.

Thanks for the reminder! I have also been eager to land this. Taking a look now.

> Also I'd like to remind you that just like catch this creates problems in browsers not compliant with ES5 allowing reserved words to be used as property names.
Assuming build process handles that and will quote it just like 'catch', it's worth noting in the readme at least.

* Yup the build handles this correctly
* I'll add an additional snippet to: https://github.com/stefanpenner/es6-promise#usage-in-ie9

**Comment by stefanpenner** [comment]
@l1bbcsg updated, waiting for CI to complete

**Pull request merged**